File: lib/data/database/daos/task_dao.dart
import 'package:drift/drift.dart';
import '../app_database.dart';
import '../tables.dart';

part 'task_dao.g.dart';

@DriftAccessor(tables: [DailyTasks])
class TaskDao extends DatabaseAccessor<AppDatabase> with _$TaskDaoMixin {
  TaskDao(super.db);

  Stream<List<DailyTaskEntity>> watchByDate(String date) =>
      (select(dailyTasks)
            ..where((t) => t.date.equals(date))
            ..orderBy([(t) => OrderingTerm(expression: t.target, mode: OrderingMode.desc)]))
          .watch();

  Future<List<DailyTaskEntity>> getByDate(String date) =>
      (select(dailyTasks)..where((t) => t.date.equals(date))).get();

  Future<void> upsertTask(DailyTasksCompanion entry) async {
    await into(dailyTasks).insertOnConflictUpdate(entry);
  }

  Future<void> updateProgress(String id, int progress, bool completed) async {
    await (update(dailyTasks)..where((t) => t.id.equals(id)))
        .write(DailyTasksCompanion(progress: Value(progress), completed: Value(completed)));
  }

  String _genId() {
    final now = DateTime.now();
    return 'task_${now.microsecondsSinceEpoch}';
  }

  Future<String> insertTask(DailyTasksCompanion entry) async {
    final id = entry.id.present ? entry.id.value : _genId();
    await into(dailyTasks).insert(entry.copyWith(id: Value(id)));
    return id;
  }
}

File: app/src/main/java/com/salesquest/sales_quest/data/dao/TaskDao.kt
package com.salesquest.sales_quest.data.dao

import androidx.room.Dao
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.Query
import androidx.room.Update
import com.salesquest.sales_quest.data.entity.DailyTaskEntity
import kotlinx.coroutines.flow.Flow

@Dao
interface TaskDao {

    @Query("SELECT * FROM daily_tasks WHERE date = :date ORDER BY target DESC")
    fun watchByDate(date: String): Flow<List<DailyTaskEntity>>

    @Query("SELECT * FROM daily_tasks WHERE date = :date")
    suspend fun getByDate(date: String): List<DailyTaskEntity>

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun upsertTask(task: DailyTaskEntity)

    @Query("UPDATE daily_tasks SET progress = :progress, completed = :completed WHERE id = :id")
    suspend fun updateProgress(id: String, progress: Int, completed: Boolean)

    @Query("DELETE FROM daily_tasks WHERE date = :date")
    suspend fun deleteByDate(date: String)

    @Query("DELETE FROM daily_tasks")
    suspend fun clearAll()
}
